### PR TITLE
Tracing

### PR DESCRIPTION
**Pull request by yogacat**
Created a graph with Vertexes and Edges. Added processing.
- [x] Optimize code so there are no duplications
- [x] Write unit tests for edge cases
- [x] Finish Main class
- [x] Write documentation

Calculates average latency for the path, shortest trace between services, traces with max number of stops etc.

### CODE REVIEW

File: src/main/java/ua/olena/tracing/graph/Graph.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    private int getAverage(Stack<String> visited) {
+        int average = 0;
+        Queue<String> queue = new ArrayDeque<>(visited);
+        String from = queue.poll();
+        while (!queue.isEmpty()) {
+            String to = queue.poll();
+            Edge edge = edges.get(from + to);
+            List<Integer> latencies = edge.getLatencies();
```
Comment (doggy0815) [inline src/main/java/ua/olena/tracing/graph/Graph.java:213]: see comment far above. there should be only one edge with one latency between two vertices to by best understanding.

File: src/main/java/ua/olena/tracing/graph/Graph.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+            depth = depthStack.pop() + 1;
+            visited.add(node);
+            Vertex vertex = vertexes.get(node);
+            Map<String, Vertex> neighbors = vertex.getNeighbors();
+
+            //If I'm going from C to C and I'm in C but this is starting point and not destination
+            if (!node.equals(to) || visited.size() == 1) {
+                for (String neighbor : neighbors.keySet()) {
+                    if (!visited.contains(neighbor) || (visited.contains(neighbor) && neighbor.equals(to))) {
+                        stack.push(neighbor);
```
Comment (doggy0815) [inline src/main/java/ua/olena/tracing/graph/Graph.java:128]: By pushing the neighbours on the stack you're creating the next iteration loop.
So if you're on "A" you have neighbours "B" and "E". Then you push  "B" and "E" on the stack to check later (unless they're in the "visited" stack.

So fine so good. 

Tree is:
```
AB5, BC4, AD5, DC8, DE6 .....

 [A] ---5---> [B] ---4---> [C] 
    \---5---> [D] ---8---> [C]
               \---6--->[E]
```

So you figure out after a while that the path A-B-C has latency "9". But you're not using this information for your next iteration steps.

You could optimise the code in that way that you evaluate the latency before pushing a vertex to the stack.
So suppose you travelled back to A-D. In this case latency between A and D is 5. 

If you now would do a look ahead for neighbours' latency then you know that next neighbour D-C has 8 and D-E has 6... So you already know A-D-C has latency 13 and A-D-E has latency 11. And both are higher than 9, so you don't need to follow this track anymore. For huge graphs this might save you some extra hops.

File: src/main/java/ua/olena/tracing/graph/Graph.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        Stack<String> stack = new Stack<>();
+        Stack<Integer> depthStack = new Stack<>();
+        stack.push(from);
+        depthStack.push(depth);
+        while (!stack.isEmpty()) {
+            String node = stack.pop();
+            depth = depthStack.pop() + 1;
+            visited.add(node);
+            Vertex vertex = vertexes.get(node);
+            Map<String, Vertex> neighbors = vertex.getNeighbors();
```
Comment (doggy0815) [inline src/main/java/ua/olena/tracing/graph/Graph.java:172]: .... aaaaand another idea. This part of your LukeGraphWalker code is more or less equal in all algorithms.
If you have passion, you could create a frame-method containing those stacks and the while loop. Within the while loop you could then call the required algorithm-method / lambda / whatever to iterate on your data

File: src/main/java/ua/olena/tracing/graph/Vertex.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    Vertex(String name) {
+        this.name = name;
+    }
+
+    String getName() {
+        return name;
+    }
+
+    void setName(String name) {
```
Comment (doggy0815) [inline src/main/java/ua/olena/tracing/graph/Vertex.java:25]: Is getter and setter ever used?

**Review by doggy0815 — APPROVED** [review]
Hey I'm done done with the review. Didn't checked the tests en detail but looks good.

File: src/main/java/ua/olena/tracing/graph/Graph.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    /**
+     * Get shortest trace between two nodes for provided graph.
+     *
+     * @param from Node to start from
+     * @param to Node to finish at
+     * @return Minimal latency of the path between two nodes
+     */
+    public Integer getShortestTrace(String from, String to) {
```
[the remark this answers is not in the document]
Comment (yogacat) [reply src/main/java/ua/olena/tracing/graph/Graph.java:108]: The idea is based on depth first traversal, but then I had to rethink some parts.

File: src/main/java/ua/olena/tracing/graph/Vertex.java (hunk 4 above)
Comment (yogacat) [reply src/main/java/ua/olena/tracing/graph/Vertex.java:25]: Yep, I'll remove the field.

File: README.md [hunk 6]
```diff
@@ -1 +1,47 @@
...
\ No newline at end of file
+# Tracing
+The application calculates 
+- average latency for the sequence of services
+- number of traces from a service to another service with maximum or exact number of hops
+- the length of shortest traces from a service to a service
+- the number of traces from a service to a service with latency less than specified number
+
+## File format
+This application processes files in CSV format. Extension of the file does not matter, file should contain lines separated by line breaks. Each line must contain service names with length of one letter for the service connection was opened from to a service it is accessing. Third parameter is a number which describes latency between these services.
```
[the remark this answers is not in the document]
Comment (yogacat) [reply README.md]: Fixed

File: src/main/java/ua/olena/tracing/Main.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package ua.olena.tracing;
+
+import ua.olena.tracing.file.FileProcessor;
+
+import static java.lang.System.out;
+
+/**
+ * Main class processing requests.
+ *
+ * @author Olena Openko, olena.openko@censhare.com
```
[the remark this answers is not in the document]
Comment (yogacat) [reply src/main/java/ua/olena/tracing/Main.java]: Well I put it because I found yours somewhere... Fixed.

File: src/main/java/ua/olena/tracing/file/FileProcessor.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+import ua.olena.tracing.graph.Graph;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static java.lang.System.*;
```
[the remark this answers is not in the document]
Comment (yogacat) [reply src/main/java/ua/olena/tracing/file/FileProcessor.java]: Fixed

File: src/main/java/ua/olena/tracing/file/FileProcessor.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            out.println(graph.getAverageLatency("A-D-C"));
+
+            //4. The average latency of the trace A-E-B-C-D.
+            out.println(graph.getAverageLatency("A-E-B-C-D"));
+
+            //5. The average latency of the trace A-E-D.
+            out.println(graph.getAverageLatency("A-E-D"));
+
+            //6. The number of traces originating in service C and ending in service C with a maximum of 3 hops.
+            out.println(graph.getTracesByStops("C", "C", 3, Condition.MAX_STOPS));
```
[the remark this answers is not in the document]
Comment (yogacat) [reply src/main/java/ua/olena/tracing/file/FileProcessor.java:48]: :P I was trying to optimise it

File: src/main/java/ua/olena/tracing/graph/Graph.java (hunk 3 above)
Comment (yogacat) [reply src/main/java/ua/olena/tracing/graph/Graph.java:172]: I thought about using lambda inside, didn't do it yet because I was afraid to make it too complex, so I won't understand it any longer

File: src/main/java/ua/olena/tracing/file/FileProcessor.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        } else {
+            throw new IllegalArgumentException("Wrong file path: " + path);
+        }
+
+    }
+
+    private Graph buildGraph(List<String> line) {
+        Graph graph = new Graph();
+        for (String trace : line) {
+            String from = trace.substring(0, 1);
```
[the remark this answers is not in the document]
Comment (yogacat) [reply src/main/java/ua/olena/tracing/file/FileProcessor.java]: Added validation to throw an exception if this happens. I cannot process these kind of traces because I don't know where one service name ends and another one starts.

File: src/main/java/ua/olena/tracing/graph/Graph.java (hunk 1 above)
Comment (yogacat) [reply src/main/java/ua/olena/tracing/graph/Graph.java:213]: I did not understand your point here. I'm getting one edge and all the traces between these two services. What did I do wrong?

File: src/main/java/ua/olena/tracing/graph/Graph.java (hunk 2 above)
Comment (yogacat) [reply src/main/java/ua/olena/tracing/graph/Graph.java:128]: Cache! Cache? I think it's cache I need here. 
I'm not sure how to easily push A-D-C calculations to cache but I can make it easier by pushing average traces to a cache (especially taken that there are not that many services in the example, however I'll have to invalidate it on adding a new edge...) Thanks!

Comment (yogacat) [reply src/main/java/ua/olena/tracing/graph/Graph.java:136]: Added cache to at least not recalculate the same thing over and over. And invalidation after adding a new edge (as it's a map, no issues I expect there)

File: src/main/java/ua/olena/tracing/graph/Graph.java (hunk 3 above)
Comment (yogacat) [reply src/main/java/ua/olena/tracing/graph/Graph.java:180]: I decided not to do it because I'm passing 10 parameters to the method, then do something inside and return nothing. Sure it can be optimized but this time I would skip it for the sake of readability.
